File: Triggers/OverlapPairSetFlagTrigger.cs
using Celeste.Mod.Entities;
using Lucky.Kits.Collections;
using LuckyHelper.Extensions;
using LuckyHelper.Modules;

namespace LuckyHelper.Triggers;

[CustomEntity("LuckyHelper/OverlapPairSetFlagTrigger")]
[Tracked]
public class OverlapPairSetFlagTrigger : Trigger
{
    private bool main;
    public string ID;
    private string flag;
    private Vector2 targetPos;

    public OverlapPairSetFlagTrigger(EntityData data, Vector2 offset) : base(data, offset)
    {
        main = data.Bool("main");
        ID = data.Attr("triggerID");
        flag = data.Attr("flag");
        targetPos = Position;
        if (data.Nodes.Length != 0)
        {
            targetPos += data.Nodes[0] + offset - Center;
        }
    }

    public override void Added(Scene scene)
    {
        base.Added(scene);
        Position = targetPos;
    }


    public override void Update()
    {
        base.Update();
        if (!main)
            return;

        bool on = false;
        foreach (OverlapPairSetFlagTrigger tri in this.GetEntities<OverlapPairSetFlagTrigger>()
                     .Where(entity => ((OverlapPairSetFlagTrigger)entity).ID == ID))
        {
            if (tri != null && tri != this && CollideCheck(tri))
            {
                on = true;
                break;
            }
        }

        this.Session().SetFlag(flag, on);
    }
}

File: ModFolder/Loenn/triggers/OverlapPairSetFlagTrigger.lua
local trigger = {}

trigger.name = "LuckyHelper/OverlapPairSetFlagTrigger"
trigger.nodeLimits = { 0, 1 }  -- 初始额外节点数, 最大额外节点数, 每个entity初始有个节点为主节点(此时有两个node)
trigger.placements = {
    name = "normal",
    data = {
        main = true,
        triggerID = "1",
        flag = ""
    }
}

trigger.fieldInformation = {
}

return trigger
